events/janus_wsevh.c

```diff
@@ -575,60 +575,57 @@ json_t *janus_wsevh_handle_request(json_t *request) {
 		}
 }
 
-/* Thread to implement the WebSockets loop */
-static void *janus_wsevh_thread(void *data) {
-	JANUS_LOG(LOG_VERB, "Joining WebSocketsEventHandler client thread\n");
-	while(g_atomic_int_get(&initialized) && !g_atomic_int_get(&stopping)) {
-		/* Loop until we have to stop */
-		if(!reconnect) {
-			lws_service(context, 50);
-		} else {
-			/* We should reconnect, get rid of the previous context */
-			if(reconnect_retries > 0) {
-				/* Wait a few seconds before retrying */
-				gint64 now = janus_get_monotonic_time();
-				if((now-disconnected) < (gint64)reconnect_retries*G_USEC_PER_SEC) {
-					/* Try again later */
-					g_usleep(100000);
-					continue;
+#if (LWS_LIBRARY_VERSION_MAJOR >= 3 && LWS_LIBRARY_VERSION_MINOR >= 2) || (LWS_LIBRARY_VERSION_MAJOR >= 4)
+	/*
+	* Websocket thread loop for websocket library newer thant 3.2
+	* The reconnect is handled in a dedicated lws scheduler janus_wsevh_schedule_connect_attempt
+	*/
+	static void *janus_wsevh_thread(void *data) {
+		JANUS_LOG(LOG_VERB, "Joining WebSocketsEventHandler (lws>=3.2) client thread\n");
+		int n = 0;
+		while(n >= 0 && g_atomic_int_get(&initialized) && !g_atomic_int_get(&stopping))
+				n = lws_service(context, 0);
+		lws_context_destroy(context);
+		JANUS_LOG(LOG_VERB, "Leaving WebSocketsEventHandler (lws>=3.2) client thread\n");
+		return NULL;
+	}
+#else
+	/*
+	* Websocket thread loop for websocket library prior to (less than) 3.2
+	* The reconnect is handled in the loop for lws < 3.2
+	*/
+	static void *janus_wsevh_thread(void *data) {
+		JANUS_LOG(LOG_VERB, "Joining WebSocketsEventHandler (lws<3.2) client thread\n");
+		while(g_atomic_int_get(&initialized) && !g_atomic_int_get(&stopping)) {
+			/* Loop until we have to stop */
+			if(!reconnect) {
+				lws_service(context, 50);
+			} else {
+				/* We should reconnect, get rid of the previous context */
+				if(reconnect_delay > 0) {
+					/* Wait a few seconds before retrying */
+					gint64 now = janus_get_monotonic_time();
+					if((now-disconnected) < (gint64)reconnect_delay*G_USEC_PER_SEC) {
+						/* Try again later */
+						g_usleep(100000);
+						continue;
+					}
+				}
+				ws_client = NULL;
+				janus_wsevh_connect_attempt(NULL);
+				if (!wsi) {
+					janus_wsevh_calculate_reconnect_delay_on_fail();
+					disconnected = janus_get_monotonic_time();
+					reconnect = TRUE;
```

Comment:
I guess it is ok to set `disconnected` and `reconnect` inside `janus_wsevh_connect_attempt`, just keeping the delay calculation and the log here.

Reply:
inside the connect attempt i set the reconnect to false as i am currently inbetween the connect and the result.
The main interaction is the callback that gives me the final async result of the operation.
It looks like there were old wsl implementations that used to give a synchronous connection feedback of the connect attempt but i never saw that. the result object is always filled

so i would actually prefer to get rid of that pointer check as to me the handling should be connect initiates something, the event tells me what happend and how to proceed.

(maybe thats grown code, it does not make sense to me... i just adopted it the way it was, the null result somehow was at some point associated with failed but i never saw it happening that way)

Reply:
I adopted it but i guess the code will anyway never run and should be removed as everything comes through the callback anyway.